src/misc.js

```diff
@@ -37,6 +37,33 @@ engine.traceFn = function (x, label, expr) {
   return x;
 };
 
+/**
+ * Used by the `defineVariable` function to inject a variable into the context
+ * @param {*} x The context data to use as the variable (if the expression is not provided)
+ * @param {*} label The name of the variable to define/set
+ * @param {*} expr (optional) An expression to run on the x parameter
+ * @returns x (The function should appear to be transparent to the caller)
+ */
+engine.defineVariable = function (x, label, expr) {
+  let data = x;
+  if (expr){
+    data = expr(x);
+  }
+  // console.log("setting variable", label);
+  // console.log("Injecting variable :[" + (label) + "]", JSON.stringify(data, null, " "));
+
+  // Just in time initialization of definedVars
+  if (!this.definedVars) this.definedVars = {};
+
+  if (Object.keys(this.definedVars).includes(label)) {
+    // console.log("Overwriting variable", label);
+    throw new Error("Variable %" + label + " already defined");
+  }
+
```

Comment:
Are defined variables allowed to override environment variables?

Reply:
Yes as per the current fhirpath spec.
https://build.fhir.org/ig/HL7/FHIRPath/index.html#definevariablename-string--expr-expression
> If the name already exists in the current expression scope, the evaluation will end and signal an error to the calling environment.